keps/sig-network/4974-deprecate-endpoints/README.md

```diff
@@ -0,0 +1,641 @@
+indicating that they are deprecated as of 1.33.
+
+We will add `APILifecycleDeprecated()` and `APILifecycleReplacement()`
+methods to `v1.Endpoints` as follows:
+
+```golang
+func (in *Endpoints) APILifecycleDeprecated() (major, minor int) {
+	return 1, 33
+}
+
+func (in *Endpoints) APILifecycleReplacement() schema.GroupVersionKind {
+	return schema.GroupVersionKind{Group: "discovery.k8s.io", Version: "v1", Kind: "EndpointSlice"}
+}
+```
+
+This will cause all operations on Endpoints objects to return the
+warning:
+
+```text
+v1 Endpoints is deprecated in v1.33+; use discovery.k8s.io/v1 EndpointSlice
+```
+
+### Warnings via metrics
+
+We will add a metric to the apiserver counting the number of Endpoints
+API operations, labeled by service account name, to help
+administrators find and update clients that are still using Endpoints.
+
+To avoid cardinality problems, we will only add a fixed number of
+labels, and just list further clients as something like "other" or
+"additional clients".
+
+### Documentation Updates
+
+A few examples in the official documentation still need to be updated
+to use EndpointSlices rather than Endpoints.
+
+### `kubernetes.default` Endpoints
+
+Both the Endpoints and the EndpointSlice for the `kubernetes.default`
+Service are created by the apiserver rather than by
+kube-controller-manager (and are thus independent of whether the
+Endpoints controller is running or not).
+
+For now, we will not change this (beyond that anyone reading the
+object will now see the deprecation warning).
+
+If, in the future, we decide to disable the Endpoints controller by
+default, we can consider whether it makes sense to stop creating the
+apiserver Endpoints as well. It is possible that we could decide that
+there is a stronger "API" guarantee around the existence of the
+`kubernetes` Endpoints object than there is around Endpoints objects
+in general.
+
+### Endpoints Cleanup
+
+We do not want to leave stale `Endpoints` objects around forever if
+the Endpoints controllers are not running. (This is both a waste of
+disk space and a potential source of confusion since the Endpoints
+objects would quickly become out-of-date and incorrect.)
+
+One possibility would be to just document that administrators should
+delete all existing Endpoints themselves if they are going to disable
+the controllers.
+
+Another possibility would be to create an `endpoints-cleanup`
+controller, that could be enabled explicitly, and document that admins
+should (probably) enable that controller if they disable the others.
+
+Or perhaps the EndpointSlice controller could delete Endpoints objects
+that were more than 24 hours out of date with respect to their
+EndpointSlices?
+
+In all cases, we should probably not automatically delete Endpoints
+that don't looke like they were originally created by the Endpoints
+controller. (That is, we should not delete Endpoints unless they
+correspond to a Service with a selector.)
```

Comment:
What if the Service was deleted in the meantime?

I think we never decided to create any annotation or use OwnerRef to reference the service, so we don't have a good mechanism to discover that programatically? Should we add it now for the purpose of doing that discovery?

Reply:
(nit: a label seems like a better fit than an annotation)

Reply:
So note that in that case we _already_ end up leaving an orphan Endpoints around.

Various reviewers have suggested that we should start having the Endpoints controller label its creations (eg, https://github.com/kubernetes/enhancements/pull/4975#discussion_r1945686819).